src/parthenon_manager.cpp

```diff
@@ -209,69 +224,90 @@ void ParthenonManager::RestartPackages(Mesh &rm, RestartReader &resfile) {
 
   size_t nCells = bsize[0] * bsize[1] * bsize[2];
 
-  // Get list of variables, assumed same for all blocks
-
-  // TODO(JL) this won't work when reading true sparse variables, will be updated in PR
-  // #383
+  // Get list of variables, they are the same for all blocks (since all blocks have the
+  // same variable metadata)
   const auto indep_restart_vars =
       mb.meshblock_data.Get()
           ->GetVariablesByFlag(
               {parthenon::Metadata::Independent, parthenon::Metadata::Restart}, false)
           .vars();
 
+  const auto sparse_info = resfile.GetSparseInfo();
+  // create map of sparse field labels to index in the SparseInfo table
+  std::unordered_map<std::string, int> sparse_idxs;
+  for (int i = 0; i < sparse_info.num_sparse; ++i) {
+    sparse_idxs.insert({sparse_info.labels[i], i});
+  }
+
   // Allocate space based on largest vector
-  size_t vlen = 1;
-  for (auto &v : indep_restart_vars) {
-    if (v->GetDim(4) > vlen) {
-      vlen = v->GetDim(4);
+  int max_vlen = 1;
+  int num_sparse = 0;
+  for (auto &v_info : indep_restart_vars) {
+    const auto &label = v_info->label();
+
+    // check that variable is in the list of sparse fields if and only if it is sparse
+    if (v_info->IsSparse()) {
+      ++num_sparse;
+      PARTHENON_REQUIRE_THROWS(sparse_idxs.count(label) == 1,
+                               "Sparse field " + label +
+                                   " is not marked as sparse in restart file");
+    } else {
+      PARTHENON_REQUIRE_THROWS(sparse_idxs.count(label) == 0,
+                               "Dense field " + label +
+                                   " is marked as sparse in restart file");
```

Comment:
This could be an interesting option (for the futures), i.e., allow restart where the originally sparse variables are fully dense.
I could imagine a use case for this (assuming sparse has a measurable overhead when it's effectively dense), when a field is initially concentrated by covers more and more of the domain over time.